File: LC735.cpp
// problem: 735. Asteroid Collision
// difficulty: medium
// runtime: 0ms, beats 100%
// memory: 76%

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> arr(asteroids.size());
        int last=-1;

        for (int i : asteroids) {
            if (i > 0) {
                arr[++last] = i;
            } else {
                if (last == -1) arr[++last] = i;
                else {
                    i = abs(i);

                    while (last != -1) {
                        int top = arr[last];

                        if (top > 0) {
                            if (i > top) {
                                last--;

                                if (last == -1) {
                                    arr[++last] = -i;
                                    break;
                                }
                            } else if (i < top) {
                                break;
                            } else {
                                last--;
                                break;
                            }
                        } else {
                            arr[++last] = -i;
                            break;
                        }
                    }
                }
            }
        }

        if (last == -1) return vector<int>();
        
        arr.resize(last+1);
        return arr;
    }
};

File: LC735.java
// problem: 735. Asteroid Collision
// difficulty: medium
// runtime: 1ms, beats 100%
// memory: 22%
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int[] ward = new int[asteroids.length];
        int last = -1;

        for (int i : asteroids) {
            if (i > 0) {
                ward[++last] = i;
            } else {
                if (last == -1) ward[++last] = i;
                else {
                    i = Math.abs(i);

                    while (last != -1) {
                        int peek = ward[last];

                        if (peek > 0) {
                            if (i > peek) {
                                last--;
                                                             
                                if (last == -1) {
                                    ward[++last] = -i;
                                    break;
                                }
                            } else if (i < peek) {
                                break;
                            } else if (i == peek) {
                                last--;
                                break;
                            } 
                        } else {
                            ward[++last] = -i;
                            break;
                        }
                    }
                }
            }
        }

        if (last == -1) return new int[0];

        int[] aste = new int[last+1];

        while (last > -1) {
            aste[last] = ward[last--];
        }

        return aste;
    }
}
